velox/functions/prestosql/ArrayDuplicates.cpp

```diff
@@ -0,0 +1,242 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <folly/container/F14Map.h>
+
+#include "velox/expression/EvalCtx.h"
+#include "velox/expression/Expr.h"
+#include "velox/expression/VectorFunction.h"
+#include "velox/functions/lib/LambdaFunctionUtil.h"
+#include "velox/functions/sparksql/Comparisons.h"
+
+namespace facebook::velox::functions {
+namespace {
+
+// See documentation at https://prestodb.io/docs/current/functions/array.html
+///
+/// This class implements the array_duplicates query function.
+///
+/// Along with the hash map, we maintain a `hasNull` flag that indicates
+/// whether null is present in the array.
+///
+/// Zero element copy:
+///
+/// In order to prevent copies of array elements, the function reuses the
+/// internal elements() vector from the original ArrayVector.
+///
+/// First a new vector is created containing the indices of the elements
+/// which will be present in the output, and wrapped into a DictionaryVector.
+/// Next the `lengths` and `offsets` vectors that control where output arrays
+/// start and end are wrapped into the output ArrayVector.
+template <typename T>
+class ArrayDuplicatesFunction : public exec::VectorFunction {
+ public:
+  // Execute function.
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      exec::Expr* caller,
+      exec::EvalCtx* context,
+      VectorPtr* result) const override {
+    // Acquire the array elements vector.
+    auto arrayVector = args.front()->as<ArrayVector>();
+    auto elementsVector = arrayVector->elements();
+    ArrayDuplicatesFunction::validateElementType(elementsVector->typeKind());
+
+    auto elementsRows =
+        toElementRows(elementsVector->size(), rows, arrayVector);
+    exec::LocalDecodedVector elements(context, *elementsVector, elementsRows);
+
+    vector_size_t elementsCount = elementsRows.size();
+    vector_size_t rowCount = arrayVector->size();
+
+    // Allocate new vectors for indices, length and offsets.
+    memory::MemoryPool* pool = context->pool();
+    BufferPtr newIndices =
+        AlignedBuffer::allocate<vector_size_t>(elementsCount, pool);
+    BufferPtr newLengths =
+        AlignedBuffer::allocate<vector_size_t>(rowCount, pool);
+    BufferPtr newOffsets =
+        AlignedBuffer::allocate<vector_size_t>(rowCount, pool);
+
+    // Pointers and cursors to the raw data.
+    vector_size_t indicesCursor = 0;
+    auto* rawNewIndices = newIndices->asMutable<vector_size_t>();
+    auto* rawSizes = newLengths->asMutable<vector_size_t>();
```

Comment:
inconsistent naming: rawNewIndices vs. rawSizes; either rename rawNewIndices to rawIndices or rename rawSizes/Offsets to rawNewSizes/Offsets